packages/visual-editing/src/util/dragAndDrop.ts

```diff
@@ -0,0 +1,308 @@
+    }
+
+    const rayRight = {
+      x1: origin.x,
+      y1: origin.y,
+      x2: origin.x + 100_000_000,
+      y2: origin.y,
+    }
+
+    return {
+      left: findClosestIntersection(rayLeft, targets),
+      right: findClosestIntersection(rayRight, targets),
+    }
+  } else {
+    const rayTop = {
+      x1: origin.x,
+      y1: origin.y,
+      x2: origin.x,
+      y2: origin.y - 100_000_000,
+    }
+
+    const rayBottom = {
+      x1: origin.x,
+      y1: origin.y,
+      x2: origin.x,
+      y2: origin.y + 100_000_000,
+    }
+
+    return {
+      top: findClosestIntersection(rayTop, targets),
+      bottom: findClosestIntersection(rayBottom, targets),
+    }
+  }
+}
+
+function findRectSanityData(rect: OverlayRect, overlayGroup: OverlayElement[]) {
+  return overlayGroup.find((e) => rectEqual(getRect(e.elements.measureElement), rect))
+    ?.sanity as SanityNode
+}
+
+function resolveInsertMsg(
+  overlayGroup: OverlayElement[],
+  insertPosition: DragInsertPosition,
+  flow: string,
+): OverlayMsgUpdateDragInsertPosition {
+  if (Object.values(insertPosition).every((v) => v === null))
+    return {
+      type: 'overlay/updateDragInsertPosition',
+      insertPosition: null,
+    }
+
+  if (flow === 'horizontal') {
+    return {
+      type: 'overlay/updateDragInsertPosition',
+      insertPosition: {
+        left: insertPosition.left
+          ? {
+              rect: insertPosition.left,
+              sanity: findRectSanityData(insertPosition.left, overlayGroup),
+            }
+          : null,
+        right: insertPosition.right
+          ? {
+              rect: insertPosition.right,
+              sanity: findRectSanityData(insertPosition.right, overlayGroup),
+            }
+          : null,
+      },
+    }
+  } else {
+    return {
+      type: 'overlay/updateDragInsertPosition',
+      insertPosition: {
+        top: insertPosition.top
+          ? {
+              rect: insertPosition.top,
+              sanity: findRectSanityData(insertPosition.top, overlayGroup),
+            }
+          : null,
+        bottom: insertPosition.bottom
+          ? {
+              rect: insertPosition.bottom,
+              sanity: findRectSanityData(insertPosition.bottom, overlayGroup),
+            }
+          : null,
+      },
+    }
+  }
+}
+
+export function handleOverlayDrag(
+  overlayGroup: OverlayElement[],
+  dragState: DragState,
+  handler: OverlayEventHandler,
+): void {
+  const rects = overlayGroup.map((e) => getRect(e.elements.measureElement))
+  const flow = calcTargetFlow(rects)
+
+  const handleMouseMove = (e: MouseEvent): void => {
+    const mousePos = {
+      x: e.clientX,
+      y: e.clientY + window.scrollY,
+    }
+
+    handler({
+      type: 'overlay/updateDragCursorPosition',
+      x: mousePos.x,
+      y: mousePos.y,
+    })
+
+    const insertPosition = calcInsertPosition(mousePos, rects, flow)
+
+    if (JSON.stringify(insertPosition) !== JSON.stringify(dragState.insertPosition)) {
```

Comment:
this could be updated later to avoid using `JSON.stringify`

Reply:
As long as we're not sensitive to object key ordering might change, or array values, I think `JSON.stringify` is the most efficient way to compare them anyway?

Reply:
I naively assumed comparing the numeric values would be more efficient, but a quick benchmark suggests `JSON.stringify` is faster! 🎉 Also, key ordering shouldn't be an issue in this case so I'll leave this be 🚀